### PR TITLE
:sparkles: Implement code transitions

### PR DESCRIPTION
**Pull request by aarthificial**
Implements the foundation for code transitions without the nice helper methods yet.
The idea is to describe transitions using an array of code fragments together with the corresponding progress of the transition in a structure called `CodeScope`:
```ts
{
  progress: 0.5,
  fragments: [
    { // adding code:
      before: '',
      after: 'const a = 1;',
    },
    { // removing code:
      before: 'const b = 2;',
      after: '',
    },
    { // modifying code:
      before: 'const c = 3;',
      after: 'const d = 4;',
    },
    // retaining the same code:
    'const e = 5;',
  ],
}
```
A valid fragment can be either a string, a `before`/`after` pair, another code scope, or a function returning any of those (a signal value). In the final version, these scopes will be constructed by helper methods.
The basic recipe for tweening would be:
1. Construct the fragments (diffing algorithm, `edit` method, `append`, etc)
2. Set the progress to `0`.
3. Tween the progress towards `1`.
4. Replace the fragments with the simplest representation. (replace `before`/`after` pairs with strings, merge fragments, etc.)

The ability to nest code scopes within each other will allow code signals to animate independently.

Certain limitations apply:
- No ability to move code. Impossible to use with patience diff plus, only the base one.
- A nested code scope cannot be tweened while the parent scope is tweened as well. When the parent scope starts being tweened it takes a snapshot of all its fragments so any scopes nested within will stop animating.

Unlike `code-fns`, this solution uses relative positioning for animations. Instead of tweening the x and y position we tween the height (number of new lines in the fragment) and the end column (number of characters in the last line). This makes the transition relative to the code scope and lets us calculate them independently.

### CODE REVIEW

File: test/src/scenes/test.tsx [hunk 1]
```diff
@@ -277,8 +300,9 @@ export default makeScene2D(function* (view) {
 
   yield* waitFor(2);
 
-  yield txt().text('Color Themes', 1);
-  yield* c().style(materialHighlightStyle, 1);
+  // TODO: Reimplement via CodeHighlighter
```
Comment (hhenrichsen) [inline test/src/scenes/test.tsx:303]: Can we try swapping the highlighter here, even if it's just another Lezer highlighter with the other theme? It sounded like that was supported based on our discussion in Discord.

This can be another PR / issue if need be.

**Review by hhenrichsen — COMMENTED** [review]
Overall looks good, a couple changes / questions but this is a good step towards usability, I think.

File: src/Code.tsx [hunk 2]
```diff
@@ -1,352 +1,110 @@
...
-        const rule = style.module
-          .getRules()
-          .split('\n')
-          .find(rule => rule.includes(classList));
-        const color = rule.split('color:')[1]?.split(';')[0].trim();
-
-        if (!color) {
-          useLogger().warn(`Unknown theme class '${classList}'`);
-        }
+  public readonly highlighter?: CodeHighlighter;
```
[the remark this answers is not in the document]
Comment (hhenrichsen) [reply src/Code.tsx:64]: Also, will a Code node work at all if it doesn't have a highlighter?

[the remark this answers is not in the document]
Comment (aarthificial) [reply src/Code.tsx]: I set the initial value to a shared `LezerHighlighter`
Yes, the node will work without a highlighter, the code will be rendered using the `fill` color in that case.
I planned to add tweening a bit later in a separate PR

File: test/src/scenes/test.tsx [hunk 3]
```diff
@@ -214,46 +218,65 @@ export default makeScene2D(function* (view) {
...
   yield txt().text('CJK and UTF-8 Characters', 1);
   yield* c().code(
     `
-                            // Note that these do not work in GitHub Actions
-                            // where this gif is rendered.
-                            // 私🦀です
-                            function crab() {
-                              console.log('🦀')
-                            }`,
+// Note that these do not work in GitHub Actions
```
[the remark this answers is not in the document]
Comment (aarthificial) [reply test/src/scenes/test.tsx]: Good point, I reverted it back

File: test/src/scenes/test.tsx (hunk 1 above)
Comment (aarthificial) [reply test/src/scenes/test.tsx]: Done

File: src/CodeScope.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      fragments: value,
+    };
+  }
+
+  return value;
+}
+
+export function resolveScope(
+  scope: CodeScope,
+  predicate: ((scope: CodeScope) => boolean) | boolean,
```
[the remark this answers is not in the document]
Comment (aarthificial) [reply src/CodeScope.ts]: Changed to `isAfter`

File: src/CodeToken.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+export interface CodeToken {
```
[the remark this answers is not in the document]
Comment (aarthificial) [reply src/CodeToken.ts]: Changed to `CodeMetrics`

File: src/LezerHighlighter.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  colorLookup: Map<number, string>;
+}
+
+export class LezerHighlighter implements CodeHighlighter<LezerCache | null> {
+  private static classRegex = /\.(\S+).*color:([^;]+)/;
+  private readonly classLookup = new Map<string, string>();
+
+  public constructor(
+    private readonly style: HighlightStyle,
+    private readonly parserMap: Map<string, Parser> = new Map([
```
[the remark this answers is not in the document]
Comment (aarthificial) [reply src/LezerHighlighter.ts:20]: Agree, I'd leave it as a TODO for now

File: src/CodeCursor.ts [hunk 7]
```diff
@@ -236,13 +236,13 @@ export class CodeCursor {
     this.context.restore();
   }
 
-  private calculateWidth(token: CodeToken): number {
```
Comment (hhenrichsen) [inline src/CodeCursor.ts]: nit: Parameter is still named `token`

File: src/CodeFragment.ts [hunk 8]
```diff
@@ -1,17 +1,21 @@
...
 
 export type PossibleCodeFragment =
   | CodeFragment
-  | CodeToken
+  | CodeMetrics
   | {before: string; after: string}
   | string;
 
-export function tokenToFragment(value: CodeToken): CodeFragment {
+export function tokenToFragment(value: CodeMetrics): CodeFragment {
```
Comment (hhenrichsen) [inline src/CodeFragment.ts:18]: this should probably be `metricsToFragment` as well

File: src/CodeCursor.ts (hunk 7 above)
Comment (aarthificial) [reply src/CodeCursor.ts]: Done

File: src/CodeFragment.ts (hunk 8 above)
Comment (aarthificial) [reply src/CodeFragment.ts]: Done
